File: composable_agents/dsl.py
"""The authoring DSL (blueprint §3.1).

Thin sugar that emits core IR :class:`~composable_agents.ir.Node` trees. Every
combinator returns a ``Node``; nothing here does IO. Node ids are provisional
(a readable per-process counter) and get reassigned to deterministic paths by
:func:`composable_agents.freeze.freeze`, so reusing a sub-flow in two places is
fine — the freeze round-trip unshares it.

Shapes are *derived*, never declared: ``parallel`` is Dataflow because ``par`` is,
``critique`` is Feedback because ``iter_up_to`` is, and so on. The DSL can't lie
about cost.
"""

from __future__ import annotations

import itertools
from typing import Optional, Union

from .ir import (
    Ann,
    CallStep,
    ContextPolicy,
    McpTool,
    NativeTool,
    Node,
    SubContract,
    SubStep,
    ThinkStep,
)
from .kinds import ContextScope, Op, Shape, SummaryPolicy

_counter = itertools.count()


def _nid(tag: str) -> str:
    return f"{tag}#{next(_counter)}"


CtxArg = Union[ContextPolicy, ContextScope, None]


def _ctx(c: CtxArg) -> Optional[ContextPolicy]:
    if c is None:
        return None
    if isinstance(c, ContextScope):
        return ContextPolicy(scope=c)
    return c


# --------------------------------------------------------------------------- #
# Leaves.
# --------------------------------------------------------------------------- #
def call(name: str, *, ctx: CtxArg = None, ann: Optional[Ann] = None) -> Node:
    """A native HTTP hand we own."""
    return Node(op=Op.PRIM, id=_nid("call"), ann=ann,
                step=CallStep(tool=NativeTool(name), ctx=_ctx(ctx)))


def mcp_call(server: str, tool: str, *, ctx: CtxArg = None, ann: Optional[Ann] = None) -> Node:
    """A tool from an MCP server (resolved + frozen at run start)."""
    return Node(op=Op.PRIM, id=_nid("mcp"), ann=ann,
                step=CallStep(tool=McpTool(server=server, tool=tool), ctx=_ctx(ctx)))


def think(brain: str, *, ctx: CtxArg = None, ann: Optional[Ann] = None) -> Node:
    """A single model call against a named brain config."""
    return Node(op=Op.PRIM, id=_nid("think"), ann=ann,
                step=ThinkStep(brain=brain, ctx=_ctx(ctx)))


def brain_from_ctx(path: str, *, ctx: CtxArg = None) -> Node:
    """A think leaf backed by a dotctx prompt directory.

    The brain id is the dotctx path; the full settings.yaml -> Brain mapping is
    done by :mod:`composable_agents.dotctx` at deploy. Context reading is always
    explicit, never ambient.
    """
    return think(path, ctx=_ctx(ctx) or ContextPolicy(scope=ContextScope.LOCAL))


def ident() -> Node:
    return Node(op=Op.IDENT, id=_nid("ident"))


def arr(pure_name: str) -> Node:
    """Lift a registered pure function into a flow leaf."""
    return Node(op=Op.ARR, id=_nid("arr"), pure=pure_name)


def subagent(ref: str, contract: SubContract,
             *, summary_policy: Optional[SummaryPolicy] = None) -> Node:
    """An opaque child workflow carrying its contract across the firewall."""
    if summary_policy is not None and contract.summary_policy is None:
        contract = SubContract(shape=contract.shape, summary_policy=summary_policy)
    return Node(op=Op.PRIM, id=_nid("sub"), step=SubStep(ref=ref, contract=contract))


# --------------------------------------------------------------------------- #
# Structural combinators.
# --------------------------------------------------------------------------- #
def _binary(op: Op, tag: str, flows: tuple[Node, ...]) -> Node:
    if not flows:
        raise ValueError(f"{tag} needs at least one flow")
    if len(flows) == 1:
        return flows[0]
    acc = flows[0]
    for nxt in flows[1:]:
        acc = Node(op=op, id=_nid(tag), left=acc, right=nxt)
    return acc


def pipeline(*flows: Node) -> Node:
    """Run flows in sequence, threading output -> input (Pipeline)."""
    return _binary(Op.SEQ, "seq", flows)


def parallel(*flows: Node) -> Node:
    """Run flows concurrently on the same input, then merge (Dataflow)."""
    return _binary(Op.PAR, "par", flows)


def fanout(*flows: Node) -> Node:
    """Fan one input out to N branches concurrently (Dataflow).

    Same IR as :func:`parallel`; the name documents intent (homogeneous fan-out
    vs heterogeneous join). The harness collects branch results into a list.
    """
    return _binary(Op.PAR, "par", flows)


def route(pred: str, if_true: Node, if_false: Node) -> Node:
    """Choose a branch by a registered pure predicate (Branching).

    ``pred(input)`` truthy -> ``if_true``, else ``if_false``. For a model-judged
    route, put a ``think`` before a ``route`` over its (pure) verdict so the
    workflow stays deterministic.
    """
    return Node(op=Op.ALT, id=_nid("alt"), pure=pred, left=if_true, right=if_false)


def critique(bound: int, body: Node, *, until: Optional[str] = None) -> Node:
    """Iterate ``body`` up to ``bound`` times (Feedback).

    ``until`` names a registered convergence predicate; when it returns truthy
    the loop stops early. Omit it to always run ``bound`` rounds.
    """
    return Node(op=Op.ITER_UP_TO, id=_nid("iter"), bound=bound, body=body, pure=until)


def stage(planner: str) -> Node:
    """Stage a model-generated plan (Staged).

    ``planner`` is a brain/dotctx that emits a Plan. At runtime the plan is
    compiled, validated (<= Feedback, granted tools only) and run as ordinary IR.
    """
    return Node(op=Op.EVAL_PLAN, id=_nid("stage"), controller=planner)


def escalate(controller: str) -> Node:
    """Open-ended controller loop (Agent — the top of the lattice; use sparingly)."""
    return Node(op=Op.APP, id=_nid("app"), controller=controller)


# --------------------------------------------------------------------------- #
# Contract helper for Sub.
# --------------------------------------------------------------------------- #
class Contract:
    """Construct :class:`SubContract` values for ``subagent``."""

    @staticmethod
    def of(shape: Shape, summary_policy: Optional[SummaryPolicy] = None) -> SubContract:
        return SubContract(shape=shape, summary_policy=summary_policy)

    @staticmethod
    def pipeline(summary_policy: Optional[SummaryPolicy] = None) -> SubContract:
        return SubContract(Shape.PIPELINE, summary_policy)

    @staticmethod
    def dataflow(summary_policy: Optional[SummaryPolicy] = None) -> SubContract:
        return SubContract(Shape.DATAFLOW, summary_policy)

    @staticmethod
    def feedback(summary_policy: Optional[SummaryPolicy] = None) -> SubContract:
        return SubContract(Shape.FEEDBACK, summary_policy)

    @staticmethod
    def staged(summary_policy: Optional[SummaryPolicy] = None) -> SubContract:
        return SubContract(Shape.STAGED, summary_policy)

    @staticmethod
    def agent(summary_policy: Optional[SummaryPolicy] = SummaryPolicy.RESULT_ONLY) -> SubContract:
        return SubContract(Shape.AGENT, summary_policy)

File: composable_agents_alt_ts/packages/core/src/dsl.ts
import type { ContextPolicy, JSONSchema, Node, NodeAnn, Shape, SubContract, SummaryPolicy, ToolRef } from './types.js';
import { shortHash } from './hash.js';

let nextIdCounter = 0;

export interface NodeOpts {
  id?: string;
  name?: string;
  ctx?: ContextPolicy;
  inputSchema?: JSONSchema;
  outputSchema?: JSONSchema;
  timeout?: number;
  timeoutMs?: number;
  cost?: number;
  risk?: string;
  stableIds?: boolean;
}

export interface CritiqueOpts extends NodeOpts {
  stopPure?: string;
}

export function ident(opts: NodeOpts = {}): Node {
  return base('ident', opts);
}

export function arr(pure: string, opts: NodeOpts = {}): Node {
  return { ...base('arr', opts), pure };
}

export function call(name: string, opts: NodeOpts = {}): Node {
  return prim({ kind: 'native', name }, opts);
}

export function mcpCall(server: string, tool: string, opts: NodeOpts = {}): Node {
  return prim({ kind: 'mcp', server, tool }, opts);
}

export function think(brain: string, opts: NodeOpts & { model?: string; modelSettings?: Record<string, unknown> } = {}): Node {
  const step = stripUndefined({
    kind: 'think' as const,
    brain,
    ctx: opts.ctx,
    model: opts.model,
    modelSettings: opts.modelSettings,
    promptSchema: opts.inputSchema,
    replySchema: opts.outputSchema,
    inputSchema: opts.inputSchema,
    outputSchema: opts.outputSchema
  });
  return { ...base('prim', opts), step };
}

export function brainFromCtx(path: string, opts: NodeOpts & { model?: string; modelSettings?: Record<string, unknown> } = {}): Node {
  return think(`dotctx:${path}`, opts);
}

export function subagent(ref: string, contract: SubContract, opts: NodeOpts = {}): Node {
  return {
    ...base('prim', opts),
    step: stripUndefined({
      kind: 'sub' as const,
      ref,
      contract,
      ctx: opts.ctx,
      inputSchema: opts.inputSchema ?? contract.inputSchema,
      outputSchema: opts.outputSchema ?? contract.outputSchema
    })
  };
}

export function seq(left: Node, right: Node, opts: NodeOpts | NodeAnn = {}): Node {
  return { ...base('seq', opts), left, right };
}

export function par(left: Node, right: Node, opts: NodeOpts | NodeAnn = {}): Node {
  return { ...base('par', opts), left, right };
}

export function alt(pure: string, left: Node, right: Node, opts: NodeOpts | NodeAnn = {}): Node {
  return { ...base('alt', opts), pure, left, right };
}

export function route(pure: string, left: Node, right: Node, opts: NodeOpts | NodeAnn = {}): Node {
  return alt(pure, left, right, opts);
}

export function pipeline(...nodes: Node[]): Node {
  if (nodes.length === 0) return ident();
  return nodes.reduce((a, b) => seq(a, b));
}

export function parallel(...nodes: Node[]): Node {
  if (nodes.length === 0) return ident();
  if (nodes.length === 1) return nodes[0]!;
  return nodes.reduce((a, b) => par(a, b));
}

export function fanout(...nodes: Node[]): Node {
  return parallel(...nodes);
}

export function critique(bound: number, body: Node, opts: CritiqueOpts | string = {}): Node {
  const normalized = typeof opts === 'string' ? { stopPure: opts } : opts;
  const stopPure = normalized.stopPure ?? 'critiqueConverged';
  return stripUndefined({ ...base('iter_up_to', normalized), bound, body, stopPure, pure: stopPure });
}

export function iterUpTo(bound: number, body: Node, opts: CritiqueOpts | string = {}): Node {
  return critique(bound, body, opts);
}

export function evalPlan(plan: Node, opts: NodeOpts | NodeAnn = {}): Node {
  return { ...base('eval_plan', opts), plan };
}

export function stage(controllerOrPlan: string | Node, opts: NodeOpts | NodeAnn = {}): Node {
  return typeof controllerOrPlan === 'string'
    ? { ...base('eval_plan', opts), controller: controllerOrPlan }
    : evalPlan(controllerOrPlan, opts);
}

export function escalate(controller: string, opts: NodeOpts | NodeAnn = {}): Node {
  return { ...base('app', opts), controller };
}

export const Contract = {
  pipeline(extra: Partial<SubContract> | SummaryPolicy = {}) {
    return contract('Pipeline', extra);
  },
  dataflow(extra: Partial<SubContract> | SummaryPolicy = {}) {
    return contract('Dataflow', extra);
  },
  branching(extra: Partial<SubContract> | SummaryPolicy = {}) {
    return contract('Branching', extra);
  },
  feedback(extra: Partial<SubContract> | SummaryPolicy = {}) {
    return contract('Feedback', extra);
  },
  staged(extra: Partial<SubContract> | SummaryPolicy = {}) {
    return contract('Staged', extra);
  },
  agent(extra: Partial<SubContract> | SummaryPolicy = {}) {
    return contract('Agent', extra);
  },
  of(shape: Shape, extra: Partial<SubContract> | SummaryPolicy = {}) {
    return contract(shape, extra);
  }
};

function contract(shape: Shape, extra: Partial<SubContract> | SummaryPolicy): SubContract {
  if (typeof extra === 'string') return { shape, summaryPolicy: extra };
  return { shape, ...extra };
}

function prim(tool: ToolRef, opts: NodeOpts): Node {
  const step = stripUndefined({ kind: 'call' as const, tool, ctx: opts.ctx, inputSchema: opts.inputSchema, outputSchema: opts.outputSchema });
  return { ...base('prim', opts), step };
}

function base(op: Node['op'], opts: NodeOpts | NodeAnn): Node {
  const hasOptsFields = 'ctx' in opts || 'stableIds' in opts || 'timeoutMs' in opts || 'timeout' in opts || 'inputSchema' in opts || 'outputSchema' in opts || 'id' in opts;
  const nodeOpts = opts as NodeOpts;
  const annInput = hasOptsFields
    ? stripUndefined({
        name: nodeOpts.name,
        timeout: nodeOpts.timeout,
        timeoutMs: nodeOpts.timeoutMs,
        cost: nodeOpts.cost,
        risk: nodeOpts.risk,
        inputSchema: nodeOpts.inputSchema,
        outputSchema: nodeOpts.outputSchema
      })
    : (opts as NodeAnn);

  const ann = Object.keys(annInput).length > 0 ? annInput : undefined;
  const partial = stripUndefined({ op, ann, inputSchema: nodeOpts.inputSchema, outputSchema: nodeOpts.outputSchema });
  return stripUndefined({ ...partial, id: nodeOpts.id ?? makeId(op, partial, nodeOpts.stableIds) }) as Node;
}

function makeId(prefix: string, fields: unknown, stable = true): string {
  if (stable) return `${prefix}_${shortHash({ prefix, fields })}`;
  nextIdCounter += 1;
  return `${prefix}_${nextIdCounter}_${shortHash({ prefix, nextIdCounter }, 8)}`;
}

function stripUndefined<T extends Record<string, unknown>>(obj: T): T {
  for (const key of Object.keys(obj)) {
    if (obj[key] === undefined) delete obj[key];
  }
  return obj;
}
